In [1]:
import re
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os
from collections import Counter
print(os.listdir("data"))

['data.csv']


In [2]:
data = pd.read_csv('data/data.csv')
data.head(5)

imdb_id  popularity     budget     revenue                original_title  \
0  tt0369610   32.985763  150000000  1513528810                Jurassic World   
1  tt1392190   28.419936  150000000   378436354            Mad Max: Fury Road   
2  tt2908446   13.112507  110000000   295238201                     Insurgent   
3  tt2488496   11.173104  200000000  2068178225  Star Wars: The Force Awakens   
4  tt2820852    9.335014  190000000  1506249360                     Furious 7   

                                                cast          director  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...       J.J. Abrams   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...         James Wan   

                         tagline  \
0              The park is open.   
1             What a Lovely Day.   
2     One Choice Can Destroy You   
3  Every generation has a story.   
4            Vengeance Hits Home   

                                            overview  runtime  \
0  Twenty-two years after the events of Jurassic ...      124   
1  An apocalyptic story set in the furthest reach...      120   
2  Beatrice Prior must confront her inner demons ...      119   
3  Thirty years after defeating the Galactic Empi...      136   
4  Deckard Shaw seeks revenge against Dominic Tor...      137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date  vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...     6/9/2015        5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...    5/13/2015        6185   
2  Summit Entertainment|Mandeville Films|Red Wago...    3/18/2015        2480   
3          Lucasfilm|Truenorth Productions|Bad Robot   12/15/2015        5292   
4  Universal Pictures|Original Film|Media Rights ...     4/1/2015        2947   

   vote_average  release_year  
0           6.5          2015  
1           7.1          2015  
2           6.3          2015  
3           7.5          2015  
4           7.3          2015

In [3]:
len(data)

1890

# Предобработка датасета

In [4]:
answer_ls = [] # создадим список с ответами. сюда будем добавлять ответы по мере прохождения теста
# сюда можем вписать создание новых колонок в датасете

In [5]:
data['profit'] = data.revenue - data.budget

# 1. У какого фильма из списка самый большой бюджет?
Варианты ответов:
1. The Dark Knight Rises (tt1345836)
2. Spider-Man 3 (tt0413300)
3. Avengers: Age of Ultron (tt2395427)
4. The Warrior's Way	(tt1032751)
5. Pirates of the Caribbean: On Stranger Tides (tt1298650)

In [6]:
data[data.budget == data.budget.max()]

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100   

                                        genres      production_companies  \
491  Adventure|Fantasy|Action|Western|Thriller  Boram Entertainment Inc.   

    release_date  vote_count  vote_average  release_year     profit  
491    12/2/2010          74           6.4          2010 -413912431

In [7]:
# тут вводим ваш ответ и добавлем в его список ответов (сейчас для примера стоит "1")
answer_ls.append(4)

# 2. Какой из фильмов самый длительный (в минутах)
1. The Lord of the Rings: The Return of the King	(tt0167260)
2. Gods and Generals	(tt0279111)
3. King Kong	(tt0360717)
4. Pearl Harbor	(tt0213149)
5. Alexander	(tt0346491)

In [8]:
data[data.runtime == data.runtime.max()]

imdb_id  popularity    budget   revenue     original_title  \
1158  tt0279111    0.469518  56000000  12923936  Gods and Generals   

                                                   cast           director  \
1158  Stephen Lang|Jeff Daniels|Robert Duvall|Kevin ...  Ronald F. Maxwell   

                                  tagline  \
1158  The nations heart was touched by...   

                                               overview  runtime  \
1158  The film centers mostly around the personal an...      214   

                 genres                production_companies release_date  \
1158  Drama|History|War  Turner Pictures|Antietam Filmworks    2/21/2003   

      vote_count  vote_average  release_year    profit  
1158          23           5.8          2003 -43076064

In [9]:
answer_ls.append(2)

# 3. Какой из фильмов самый короткий (в минутах)
Варианты ответов:

1. Home on the Range	tt0299172
2. The Jungle Book 2	tt0283426
3. Winnie the Pooh	tt1449283
4. Corpse Bride	tt0121164
5. Hoodwinked!	tt0443536

In [10]:
data[data.runtime == data.runtime.min()]

imdb_id  popularity    budget   revenue   original_title  \
769  tt1449283    1.425344  30000000  14460000  Winnie the Pooh   

                                                  cast  \
769  Jim Cummings|Travis Oates|Jim Cummings|Bud Luc...   

                      director   tagline  \
769  Stephen Anderson|Don Hall  Oh Pooh.   

                                              overview  runtime  \
769  During an ordinary day in Hundred Acre Wood, W...       63   

               genres                               production_companies  \
769  Animation|Family  Walt Disney Pictures|Walt Disney Animation Stu...   

    release_date  vote_count  vote_average  release_year    profit  
769    4/13/2011         174           6.8          2011 -15540000

In [11]:
answer_ls.append(3)

# 4. Средняя длительность фильма?

Варианты ответов:
1. 115
2. 110
3. 105
4. 120
5. 100


In [12]:
data.describe()

popularity        budget       revenue      runtime   vote_count  \
count  1890.000000  1.890000e+03  1.890000e+03  1890.000000  1890.000000   
mean      1.560611  5.450696e+07  1.552890e+08   109.653439   785.666667   
std       1.849471  4.931705e+07  2.146386e+08    18.013641  1068.725880   
min       0.010335  5.000000e+06  2.033165e+06    63.000000    10.000000   
25%       0.643284  2.000000e+07  3.434007e+07    97.000000   164.000000   
50%       1.065065  3.800000e+07  8.358664e+07   106.500000   389.000000   
75%       1.804537  7.200000e+07  1.782136e+08   120.000000   937.750000   
max      32.985763  4.250000e+08  2.781506e+09   214.000000  9767.000000   

       vote_average  release_year        profit  
count   1890.000000   1890.000000  1.890000e+03  
mean       6.140899   2007.861905  1.007820e+08  
std        0.764584      4.467929  1.834078e+08  
min        3.300000   2000.000000 -4.139124e+08  
25%        5.600000   2004.000000  5.000000e+06  
50%        6.100000   2008.000000  4.400020e+07  
75%        6.600000   2012.000000  1.201703e+08  
max        8.100000   2015.000000  2.544506e+09

In [13]:
# или
round(data.runtime.mean())

110

In [14]:
answer_ls.append(2)

# 5. Средняя длительность фильма по медиане?
Варианты ответов:
1. 106
2. 112
3. 101
4. 120
5. 115




In [15]:
data.describe()

popularity        budget       revenue      runtime   vote_count  \
count  1890.000000  1.890000e+03  1.890000e+03  1890.000000  1890.000000   
mean      1.560611  5.450696e+07  1.552890e+08   109.653439   785.666667   
std       1.849471  4.931705e+07  2.146386e+08    18.013641  1068.725880   
min       0.010335  5.000000e+06  2.033165e+06    63.000000    10.000000   
25%       0.643284  2.000000e+07  3.434007e+07    97.000000   164.000000   
50%       1.065065  3.800000e+07  8.358664e+07   106.500000   389.000000   
75%       1.804537  7.200000e+07  1.782136e+08   120.000000   937.750000   
max      32.985763  4.250000e+08  2.781506e+09   214.000000  9767.000000   

       vote_average  release_year        profit  
count   1890.000000   1890.000000  1.890000e+03  
mean       6.140899   2007.861905  1.007820e+08  
std        0.764584      4.467929  1.834078e+08  
min        3.300000   2000.000000 -4.139124e+08  
25%        5.600000   2004.000000  5.000000e+06  
50%        6.100000   2008.000000  4.400020e+07  
75%        6.600000   2012.000000  1.201703e+08  
max        8.100000   2015.000000  2.544506e+09

In [16]:
# или
round(data.runtime.median())

106

In [17]:
answer_ls.append(1)

# 6. Какой самый прибыльный фильм?
Варианты ответов:
1. The Avengers	tt0848228
2. Minions	tt2293640
3. Star Wars: The Force Awakens	tt2488496
4. Furious 7	tt2820852
5. Avatar	tt0499549

In [18]:

data[data.profit == data.profit.max()]

imdb_id  popularity     budget     revenue original_title  \
239  tt0499549    9.432768  237000000  2781505847         Avatar   

                                                  cast       director  \
239  Sam Worthington|Zoe Saldana|Sigourney Weaver|S...  James Cameron   

                         tagline  \
239  Enter the World of Pandora.   

                                              overview  runtime  \
239  In the 22nd century, a paraplegic Marine is di...      162   

                                       genres  \
239  Action|Adventure|Fantasy|Science Fiction   

                                  production_companies release_date  \
239  Ingenious Film Partners|Twentieth Century Fox ...   12/10/2009   

     vote_count  vote_average  release_year      profit  
239        8458           7.1          2009  2544505847

In [19]:
answer_ls.append(5)

# 7. Какой фильм самый убыточный?
Варианты ответов:
1. Supernova tt0134983
2. The Warrior's Way tt1032751
3. Flushed Away	tt0424095
4. The Adventures of Pluto Nash	tt0180052
5. The Lone Ranger	tt1210819

In [20]:
data[data.profit == data.profit.min()]

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100   

                                        genres      production_companies  \
491  Adventure|Fantasy|Action|Western|Thriller  Boram Entertainment Inc.   

    release_date  vote_count  vote_average  release_year     profit  
491    12/2/2010          74           6.4          2010 -413912431

In [21]:
answer_ls.append(2)

# 8. Сколько всего фильмов в прибыли?
Варианты ответов:
1. 1478
2. 1520
3. 1241
4. 1135
5. 1398


In [22]:
len(data[data.revenue > data.budget])

1478

In [23]:
answer_ls.append(1)

# 9. Самый прибыльный фильм в 2008 году?
Варианты ответов:
1. Madagascar: Escape 2 Africa	tt0479952
2. Iron Man	tt0371746
3. Kung Fu Panda	tt0441773
4. The Dark Knight	tt0468569
5. Mamma Mia!	tt0795421

In [24]:
maxprofit_2008 = data[(data.release_year == 2008)] 
maxprofit_2008[maxprofit_2008.profit == maxprofit_2008.profit.max()]

imdb_id  popularity     budget     revenue   original_title  \
600  tt0468569    8.466668  185000000  1001921825  The Dark Knight   

                                                  cast           director  \
600  Christian Bale|Michael Caine|Heath Ledger|Aaro...  Christopher Nolan   

             tagline                                           overview  \
600  Why So Serious?  Batman raises the stakes in his war on crime. ...   

     runtime                       genres  \
600      152  Drama|Action|Crime|Thriller   

                                  production_companies release_date  \
600  DC Comics|Legendary Pictures|Warner Bros.|Syncopy    7/16/2008   

     vote_count  vote_average  release_year     profit  
600        8432           8.1          2008  816921825

In [25]:
answer_ls.append(4)

# 10. Самый убыточный фильм за период с 2012 по 2014 (включительно)?
Варианты ответов:
1. Winter's Tale	tt1837709
2. Stolen	tt1656186
3. Broken City	tt1235522
4. Upside Down	tt1374992
5. The Lone Ranger	tt1210819


In [26]:
minprofit_2012_2014 = data[(data.release_year >= 2012) & (data.release_year <= 2014)] 
minprofit_2012_2014[minprofit_2012_2014.profit == minprofit_2012_2014.profit.min()]

imdb_id  popularity     budget   revenue   original_title  \
1246  tt1210819     1.21451  255000000  89289910  The Lone Ranger   

                                                   cast        director  \
1246  Johnny Depp|Armie Hammer|William Fichtner|Hele...  Gore Verbinski   

                      tagline  \
1246  Never Take Off the Mask   

                                               overview  runtime  \
1246  The Texas Rangers chase down a gang of outlaws...      149   

                        genres  \
1246  Action|Adventure|Western   

                                   production_companies release_date  \
1246  Walt Disney Pictures|Jerry Bruckheimer Films|I...     7/3/2013   

      vote_count  vote_average  release_year     profit  
1246        1607           6.0          2013 -165710090

In [27]:
answer_ls.append(5)

# 11. Какого жанра фильмов больше всего?
Варианты ответов:
1. Action
2. Adventure
3. Drama
4. Comedy
5. Thriller

In [30]:
pd.DataFrame(data.genres.str.split('|').tolist()).stack().value_counts()

Drama              782
Comedy             683
Thriller           597
Action             583
Adventure          416
Crime              315
Romance            308
Family             260
Science Fiction    248
Fantasy            223
Horror             176
Mystery            168
Animation          139
Music               64
History             62
War                 58
Western             20
Documentary          8
Foreign              2
dtype: int64

In [31]:
answer_ls.append(3)

# 12. Какого жанра среди прибыльных фильмов больше всего?
Варианты ответов:
1. Drama
2. Comedy
3. Action
4. Thriller
5. Adventure

In [32]:
data_profit = data[data['profit'] > 0]
pd.DataFrame(data_profit.genres.str.split('|').tolist()).stack().value_counts()

Drama              560
Comedy             551
Thriller           446
Action             444
Adventure          337
Romance            242
Crime              231
Family             226
Science Fiction    195
Fantasy            188
Horror             150
Animation          120
Mystery            119
Music               47
History             46
War                 41
Western             12
Documentary          7
dtype: int64

In [33]:
answer_ls.append(1)

# 13. Кто из режиссеров снял больше всего фильмов?
Варианты ответов:
1. Steven Spielberg
2. Ridley Scott 
3. Steven Soderbergh
4. Christopher Nolan
5. Clint Eastwood

In [34]:
data['director'].value_counts()

Steven Soderbergh       13
Clint Eastwood          12
Ridley Scott            12
Steven Spielberg        10
Shawn Levy              10
                        ..
Charles Shyer            1
Jan de Bont              1
Peter Landesman          1
Drew Goddard             1
Nicolas Winding Refn     1
Name: director, Length: 958, dtype: int64

In [35]:
answer_ls.append(3)

# 14. Кто из режиссеров снял больше всего Прибыльных фильмов?
Варианты ответов:
1. Steven Soderbergh
2. Clint Eastwood
3. Steven Spielberg
4. Ridley Scott
5. Christopher Nolan

In [36]:
data_profit['director'].value_counts()

Ridley Scott             12
Clint Eastwood           10
Steven Soderbergh        10
Steven Spielberg         10
Shawn Levy                9
                         ..
Karyn Kusama              1
Jocelyn Moorhouse         1
Sam Fell|Chris Butler     1
Eli Roth                  1
Hoyt Yeatman              1
Name: director, Length: 774, dtype: int64

In [37]:
answer_ls.append(4)

# 15. Кто из режиссеров принес больше всего прибыли?
Варианты ответов:
1. Steven Spielberg
2. Christopher Nolan
3. David Yates
4. James Cameron
5. Peter Jackson


In [38]:
data.groupby('director').sum().reset_index().sort_values('profit', ascending=False).head(10) 

director  popularity      budget     revenue  runtime  \
689      Peter Jackson   45.679675  1288000000  6490593685     1355   
213        David Yates   20.506537   775000000  4154295625      567   
147  Christopher Nolan   61.560756  1005000000  4167548502     1135   
348        J.J. Abrams   24.004784   740000000  3579169916      633   
587        Michael Bay   15.241191  1126000000  3886938960     1209   
353      James Cameron    9.686149   250000000  2798546718      254   
274   Francis Lawrence   26.625627   703000000  3179979588      747   
867   Steven Spielberg   23.081698   937000000  3386700791     1356   
478        Joss Whedon   13.582694   500000000  2924593677      284   
314     Gore Verbinski   21.563101  1155000000  3426362290     1058   

     vote_count  vote_average  release_year      profit  
689       31197          58.1         16059  5202593685  
213       13819          29.6          8037  3379295625  
147       41561          61.1         16057  3162548502  
348       14962          35.1         10054  2839169916  
587       13854          50.0         16063  2760938960  
353        8481          13.8          4012  2548546718  
274       14859          40.5         12065  2476979588  
867       12990          66.9         20064  2449700791  
478       13207          14.7          4027  2424593677  
314       14237          51.2         16048  2271362290

In [39]:
answer_ls.append(5)

# 16. Какой актер принес больше всего прибыли?
Варианты ответов:
1. Emma Watson
2. Johnny Depp
3. Michelle Rodriguez
4. Orlando Bloom
5. Rupert Grint

In [40]:
data_cast = pd.DataFrame(data.cast.str.split('|').tolist()).stack().reset_index() 
data_cast.columns = ['row1', 'row2', 'actor']
data_cast['profit_actor'] = data_cast.row1.apply(lambda x: data.iloc[x]['profit'])
data_cast.groupby(['actor'])['profit_actor'].sum().sort_values(ascending=False)

actor
Emma Watson           6666245597
Daniel Radcliffe      6514990281
Rupert Grint          6408638290
Ian McKellen          6087375777
Robert Downey Jr.     5316030161
                         ...    
Elisabeth Harnois     -111007242
Emilio Echevarría    -119180039
Kate Bosworth         -369455668
Ti Lung               -413912431
Jang Dong-gun         -413912431
Name: profit_actor, Length: 3408, dtype: int64

In [41]:
answer_ls.append(1)

# 17. Какой актер принес меньше всего прибыли в 2012 году?
Варианты ответов:
1. Nicolas Cage
2. Danny Huston
3. Kirsten Dunst
4. Jim Sturgess
5. Sami Gayle

In [42]:
data_actor_2012 = data.query('release_year == 2012')
data_cast = pd.DataFrame(data_actor_2012.cast.str.split('|').tolist()).stack().reset_index() 
data_cast.columns = ['row1', 'row2', 'actor']
data_cast['profit_actor'] = data_cast.row1.apply(lambda x: data_actor_2012.iloc[x]['profit'])
data_cast.groupby(['actor'])['profit_actor'].sum().sort_values()

actor
Kirsten Dunst         -68109207
Heidi Hawkins         -51893525
James Kidnie          -51893525
Timothy Spall         -51893525
Malin Ã…kerman        -43733509
                        ...    
Robert Downey Jr.    1299557910
Mark Ruffalo         1299557910
Chris Evans          1299557910
Denis Leary          1319460639
Chris Hemsworth      1542450773
Name: profit_actor, Length: 466, dtype: int64

In [43]:
answer_ls.append(3)

# 18. Какой актер снялся в большем количестве высокобюджетных фильмов? (в фильмах где бюджет выше среднего по данной выборке)
Варианты ответов:
1. Tom Cruise
2. Mark Wahlberg 
3. Matt Damon
4. Angelina Jolie
5. Adam Sandler

In [44]:
data_actor_budget = data.loc[data['budget'] > data['budget'].mean()]
data_cast = pd.DataFrame(data_actor_budget.cast.str.split('|').tolist()).stack().reset_index() 
data_cast.columns = ['row1', 'row2', 'actor']
data_cast['actor'].value_counts()

Matt Damon           18
Adam Sandler         17
Angelina Jolie       16
Samuel L. Jackson    15
Tom Cruise           15
                     ..
Shane West            1
Zaid Farid            1
Cole Konis            1
Jena Malone           1
Vera Farmiga          1
Name: actor, Length: 1508, dtype: int64

In [45]:
answer_ls.append(3)

# 19. В фильмах какого жанра больше всего снимался Nicolas Cage?  
Варианты ответа:
1. Drama
2. Action
3. Thriller
4. Adventure
5. Crime

In [46]:
data_nic_cage = data.loc[data['cast'].str.contains('Nicolas Cage')]
data_genres = pd.DataFrame(data_nic_cage.genres.str.split('|').tolist()).stack().reset_index()
data_genres.columns = ['row1', 'row2', 'genres']
data_genres['genres'].value_counts()

Action             17
Thriller           15
Drama              12
Crime              10
Fantasy             8
Adventure           7
Comedy              6
Science Fiction     4
Family              3
Animation           3
Mystery             3
History             2
Horror              1
War                 1
Romance             1
Name: genres, dtype: int64

In [47]:
answer_ls.append(2)

# 20. Какая студия сняла больше всего фильмов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [48]:
data_prod_companies = pd.DataFrame(data.production_companies.str.split('|')
                                   .tolist()).stack().reset_index()
data_prod_companies.columns = ['row1', 'row2', 'prod_companies']
data_prod_companies['prod_companies'].value_counts()


Universal Pictures                        173
Warner Bros.                              168
Paramount Pictures                        122
Columbia Pictures                         117
Twentieth Century Fox Film Corporation    109
                                         ... 
Pandora Pictures                            1
Miracle Pictures                            1
Etalon film                                 1
Coconut Pete Productions                    1
Le Grisbi Productions                       1
Name: prod_companies, Length: 1772, dtype: int64

In [49]:
data_prod_comp = data.production_companies.str.split('|').tolist()
prod_comp_dict = {}
prod_companies_ls = ['Warner Bros', 'Universal Pictures', 'Paramount Pictures', 
                     'Columbia Pictures', 'Twentieth Century Fox Film Corporation']
for item in data_prod_comp:
    for comp in item:
        for i in prod_companies_ls:
            if i in comp:  
                prod_comp_dict.setdefault(comp, 0)
                prod_comp_dict[comp] += 1
print(prod_comp_dict)

{'Universal Pictures': 173, 'Paramount Pictures': 122, 'Twentieth Century Fox Film Corporation': 109, 'Columbia Pictures': 117, 'Warner Bros.': 168, 'Universal Pictures International (UPI)': 1, 'Warner Bros. Animation': 5, 'Warner Bros. Interactive Entertainment': 1, 'Warner Bros. Pictures': 13, 'Columbia Pictures Corporation': 47, 'Universal Pictures Corporation': 1}


In [50]:
answer_ls.append(1)

# 21. Какая студия сняла больше всего фильмов в 2015 году?
Варианты ответа:
1. Universal Pictures
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [51]:
data_prod_comp_2015 = data.loc[data['release_year'] == 2015]
data_prod_comp_2015 = pd.DataFrame(data_prod_comp_2015.production_companies.str.split('|')\
                                   .tolist()).stack().reset_index()
data_prod_comp_2015.columns = ['row1', 'row2', 'prod_companies']
data_prod_comp_2015['prod_companies'].value_counts()

Warner Bros.                              12
Universal Pictures                        10
Twentieth Century Fox Film Corporation     8
Paramount Pictures                         7
Columbia Pictures                          7
                                          ..
Columbia TriStar                           1
3 Arts Entertainment                       1
A113                                       1
Flynn Picture Company                      1
Goldmann Pictures                          1
Name: prod_companies, Length: 246, dtype: int64

In [52]:
answer_ls.append(4)

# 22. Какая студия заработала больше всего денег в жанре комедий за все время?
Варианты ответа:
1. Warner Bros
2. Universal Pictures (Universal)
3. Columbia Pictures
4. Paramount Pictures
5. Walt Disney

In [53]:
def comedy_prodcompany(row):
    prod_companies_ls = ['Warner Bros.', 'Universal Pictures', 'Paramount Pictures', 
                     'Columbia Pictures', 'Twentieth Century Fox Film Corporation']
    if 'Comedy' in row['genres']:   
        for i in prod_companies_ls:
            if i in row['production_companies']:
                return i
    return 'undefined'    

In [54]:
data['comedy_company'] = data.apply(comedy_prodcompany, axis = 1)
data.groupby(['comedy_company'])['profit'].sum().sort_values(ascending=False)

comedy_company
undefined                                 161872551803
Universal Pictures                          8952422726
Columbia Pictures                           6750183034
Twentieth Century Fox Film Corporation      5686960294
Warner Bros.                                4370928916
Paramount Pictures                          2844975137
Name: profit, dtype: int64

In [55]:
answer_ls.append(2)

# 23. Какая студия заработала больше всего денег в 2012 году?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Columbia Pictures
4. Paramount Pictures
5. Lucasfilm

In [56]:
data_company_2012 = data.query('release_year == 2012')
comp_profit_2012 = pd.DataFrame(data_company_2012.production_companies.str.split('|').tolist()).stack().reset_index() 
comp_profit_2012.columns = ['row1', 'row2', 'company']
comp_profit_2012['profit_company'] = comp_profit_2012.row1.apply(lambda x: data_company_2012.iloc[x]['profit'])
comp_profit_2012.groupby(['company'])['profit_company'].sum().sort_values(ascending=False)

company
Columbia Pictures     2501406608
Universal Pictures    1981011579
Marvel Studios        1299557910
Warner Bros.          1258020056
Relativity Media      1032593938
                         ...    
Jouror Productions     -51893525
Onyx Films             -51893525
Transfilm              -51893525
Studio 37              -51893525
France 2 Cinéma       -82545651
Name: profit_company, Length: 236, dtype: int64

In [57]:
answer_ls.append(3)

# 24. Самый убыточный фильм от Paramount Pictures
Варианты ответа:

1. K-19: The Widowmaker tt0267626
2. Next tt0435705
3. Twisted tt0315297
4. The Love Guru tt0811138
5. The Fighter tt0964517

In [58]:
data.loc[data['production_companies'].str.contains('Paramount Pictures')].\
sort_values(['profit']).head()

imdb_id  popularity     budget   revenue        original_title  \
926   tt0267626    0.722330  100000000  35168966  K-19: The Widowmaker   
1124  tt0300556    1.047214   80000000  19480739              Timeline   
1723  tt0435705    1.068126   70000000  18211013                  Next   
1624  tt0375173    0.898601   60000000  13395939                 Alfie   
1632  tt0315297    0.767870   50000000  25195000               Twisted   

                                                   cast         director  \
926   Harrison Ford|Liam Neeson|Peter Sarsgaard|Joss...  Kathryn Bigelow   
1124  Paul Walker|Frances O'Connor|Gerard Butler|Bil...   Richard Donner   
1723  Nicolas Cage|Julianne Moore|Jessica Biel|Thoma...     Lee Tamahori   
1624  Jude Law|Marisa Tomei|Omar Epps|Jane Krakowski...    Charles Shyer   
1632  Ashley Judd|Samuel L. Jackson|Andy García|Dav...   Philip Kaufman   

                                                tagline  \
926                            Fate has found its hero.   
1124  They had to travel into the past to save the f...   
1723        If you can see the future, you can save it.   
1624   Meet a man who never met a woman he didn't love.   
1632                           Every murder has a mark.   

                                               overview  runtime  \
926   When Russia's first nuclear submarine malfunct...      138   
1124  A group of archaeological students become trap...      116   
1723  Las Vegas showroom magician Cris Johnson has a...       96   
1624  In Manhattan, the British limousine driver Alf...      103   
1632  Recently promoted and transferred to the homic...       97   

                                     genres  \
926                  Thriller|Drama|History   
1124       Action|Adventure|Science Fiction   
1723        Action|Science Fiction|Thriller   
1624                                 Comedy   
1632  Action|Adventure|Crime|Drama|Thriller   

                                   production_companies release_date  \
926   Paramount Pictures|Intermedia Films|National G...    7/19/2002   
1124  Paramount Pictures|Donners' Company|Mutual Fil...   11/26/2003   
1723  Paramount Pictures|Virtual Studios|Revolution ...    4/24/2007   
1624             Paramount Pictures|Patalex Productions   10/22/2004   
1632  Paramount Pictures|Kopelson Entertainment|Harl...    2/27/2004   

      vote_count  vote_average  release_year    profit      comedy_company  
926          146           6.0          2002 -64831034           undefined  
1124         176           5.1          2003 -60519261           undefined  
1723         499           5.9          2007 -51788987           undefined  
1624         140           5.4          2004 -46604061  Paramount Pictures  
1632          67           5.7          2004 -24805000           undefined

In [59]:
answer_ls.append(1)

# 25. Какой Самый прибыльный год (заработали больше всего)?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2002
5. 2015

In [60]:
data.groupby(['release_year'])['profit'].sum().sort_values(ascending=False)

release_year
2015    18668572378
2014    16397812953
2012    16077001687
2013    15243179791
2011    14730241341
2009    13423744372
2010    13117292530
2008    11663881990
2007    11565911801
2004     9634180720
2003     9228823312
2002     9002361487
2005     8981925558
2006     8691077320
2001     7950614865
2000     6101399805
Name: profit, dtype: int64

In [61]:
answer_ls.append(5)

# 26. Какой Самый прибыльный год для студии Warner Bros?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2010
5. 2015

In [62]:
data[data['production_companies'].str.contains('Warner Bros.')].\
groupby(['release_year'])['profit'].sum().sort_values()

release_year
2000     452631386
2006     620170743
2015     870368348
2002    1022709901
2012    1258020056
2001    1343545668
2005    1551980298
2004    1631933725
2013    1636453400
2009    1822454136
2003    1855493377
2011    1871393682
2010    1974712985
2008    2134595031
2007    2201675217
2014    2295464519
Name: profit, dtype: int64

In [63]:
answer_ls.append(1)

# 27. В каком месяце за все годы суммарно вышло больше всего фильмов?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [64]:
data['month'] = [release_date.split('/')[0] for release_date in data.release_date]
data['month'].value_counts(ascending = False)

9     227
12    191
10    186
8     161
3     156
4     149
6     147
11    146
7     142
5     140
2     135
1     110
Name: month, dtype: int64

In [65]:
answer_ls.append(4)

# 28. Сколько суммарно вышло фильмов летом? (за июнь, июль, август)
Варианты ответа:
1. 345
2. 450
3. 478
4. 523
5. 381

In [66]:
data[(data.month == '6') | (data.month == '7') | (data.month == '8')].month.value_counts().sum()

450

In [67]:
answer_ls.append(2)

# 29. Какой режисер выпускает (суммарно по годам) больше всего фильмов зимой?
Варианты ответов:
1. Steven Soderbergh
2. Christopher Nolan
3. Clint Eastwood
4. Ridley Scott
5. Peter Jackson

In [68]:
data[(data.month == '12') | (data.month == '1') | (data.month == '2')].director.value_counts()

Peter Jackson        7
Steven Soderbergh    6
Clint Eastwood       6
Martin Scorsese      4
Nancy Meyers         4
                    ..
Kelly Asbury         1
Tom Hooper           1
Carl Rinsch          1
Chris Columbus       1
Jon Turteltaub       1
Name: director, Length: 333, dtype: int64

In [69]:
answer_ls.append(5)

# 30. Какой месяц чаще всего по годам самый прибыльный?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [70]:
data.groupby(['month'])['profit'].sum().sort_values(ascending=False)

month
6     27551262500
12    26563071474
5     25039516141
11    18654155473
7     18307895020
3     13720216540
10    12994907859
4     12876005537
9     10878921041
8      9640610446
2      8854574425
1      5396885454
Name: profit, dtype: int64

In [71]:
answer_ls.append(2)

# 31. Названия фильмов какой студии в среднем самые длинные по количеству символов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [72]:
data['letter_count'] = data['original_title'].apply(lambda x: len(x))
letter_df = pd.DataFrame(data.production_companies.str.split('|').tolist()).stack().reset_index() 
letter_df.columns = ['row1', 'row2', 'company']
letter_df['letter_count'] = letter_df.row1.apply(lambda x: data.iloc[x]['letter_count'])
letter_df.groupby(['company']).letter_count.mean().sort_values()

company
Global Entertainment Group     2.0
Ixtlan Productions             2.0
XM2 Productions                2.0
Berlanti Productions           3.0
Everest Entertainment          3.0
                              ... 
Museum Canada Productions     46.0
Polsky Films                  46.0
Dos Corazones                 47.0
Jim Henson Company, The       59.0
Four By Two Productions       83.0
Name: letter_count, Length: 1772, dtype: float64

In [73]:
answer_ls.append(5)

# 32. Названия фильмов какой студии в среднем самые длинные по количеству слов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [74]:
data['words'] = data['original_title'].str.split(' ')
data['word_count'] = data['words'].apply(lambda x: len(x))
word_df = pd.DataFrame(data.production_companies.str.split('|').tolist()).stack().reset_index() 
word_df.columns = ['row1', 'row2', 'company']
word_df['word_count'] = word_df.row1.apply(lambda x: data.iloc[x]['word_count'])
word_df.groupby(['company']).word_count.mean().sort_values()

company
Great Wight Productions/ Osford Films     1.0
Blue Sea Productions                      1.0
Blue Productions                          1.0
Maple Shade Films                         1.0
Many Rivers Productions                   1.0
                                         ... 
Dos Corazones                             9.0
The Saul Zaentz Company                   9.0
Polsky Films                              9.0
Jim Henson Company, The                  10.0
Four By Two Productions                  12.0
Name: word_count, Length: 1772, dtype: float64

In [75]:
answer_ls.append(5)

# 33. Сколько разных слов используется в названиях фильмов?(без учета регистра)
Варианты ответа:
1. 6540
2. 1002
3. 2461
4. 28304
5. 3432

In [76]:
lower_words = pd.DataFrame(data.original_title.str.lower())
lower_words = pd.DataFrame(lower_words.original_title.str.split(' ').tolist()).stack().reset_index()
lower_words.columns = ['row1', 'row2', 'words']
lower_words['words'].nunique()

2462

In [77]:
# Dmitry Ivanov
total_title_list = data.original_title.apply(lambda x: x+' ').sum().lower()
unique_words = set(total_title_list.split())
len(unique_words)

2461

In [78]:
# Ekaterina Popova 
# Коллеги, прошу прощения, что ввела в заблуждение Уточню у коллег, предполагаю, что все же верно 2461,
#например вот таким способом:
words = set()
for i in data['original_title'].to_list():
    for j in i.lower().split():
        words.add(j)
print(len(words))

2461


In [79]:
# Однако, этот с виду верный код дает результат 2462
pd.DataFrame(data.original_title.apply(lambda x: [i.lower() for i in x.split(' ')]).tolist()).stack().nunique()

2462

In [80]:
answer_ls.append(3)

# 34. Какие фильмы входят в 1 процент лучших по рейтингу?
Варианты ответа:
1. Inside Out, Gone Girl, 12 Years a Slave
2. BloodRayne, The Adventures of Rocky & Bullwinkle
3. The Lord of the Rings: The Return of the King
4. 300, Lucky Number Slevin

In [81]:
data[data['vote_average'] > data.quantile(0.99, numeric_only=True)['vote_average']].\
sort_values('vote_average', ascending = False)

imdb_id  popularity     budget     revenue  \
600   tt0468569    8.466668  185000000  1001921825   
9     tt2096673    6.326804  175000000   853708609   
34    tt3170832    3.557846    6000000    35401758   
118   tt0816692   24.949134  165000000   621752480   
125   tt2084970    8.110711   14000000   233555708   
119   tt2015381   14.311205  170000000   773312399   
128   tt2267998    6.438727   61000000   369330363   
138   tt2278388    4.930820   30000000   174600318   
370   tt1375666    9.363643  160000000   825500000   
873   tt0253474    2.364204   35000000   120072577   
1082  tt0167260    7.122455   94000000  1118888979   
1184  tt0993846    4.877927  100000000   392000694   
1192  tt2024544    3.625529   20000000   187000000   
1801  tt0209144    3.095625    9000000    39723096   

                                     original_title  \
600                                 The Dark Knight   
9                                        Inside Out   
34                                             Room   
118                                    Interstellar   
125                              The Imitation Game   
119                         Guardians of the Galaxy   
128                                       Gone Girl   
138                        The Grand Budapest Hotel   
370                                       Inception   
873                                     The Pianist   
1082  The Lord of the Rings: The Return of the King   
1184                        The Wolf of Wall Street   
1192                               12 Years a Slave   
1801                                        Memento   

                                                   cast           director  \
600   Christian Bale|Michael Caine|Heath Ledger|Aaro...  Christopher Nolan   
9     Amy Poehler|Phyllis Smith|Richard Kind|Bill Ha...        Pete Docter   
34    Brie Larson|Jacob Tremblay|Joan Allen|Sean Bri...   Lenny Abrahamson   
118   Matthew McConaughey|Jessica Chastain|Anne Hath...  Christopher Nolan   
125   Benedict Cumberbatch|Keira Knightley|Matthew G...      Morten Tyldum   
119   Chris Pratt|Zoe Saldana|Dave Bautista|Vin Dies...         James Gunn   
128   Ben Affleck|Rosamund Pike|Carrie Coon|Neil Pat...      David Fincher   
138   Ralph Fiennes|Tony Revolori|F. Murray Abraham|...       Wes Anderson   
370   Leonardo DiCaprio|Joseph Gordon-Levitt|Ellen P...  Christopher Nolan   
873   Adrien Brody|Thomas Kretschmann|Frank Finlay|M...     Roman Polanski   
1082  Elijah Wood|Ian McKellen|Viggo Mortensen|Liv T...      Peter Jackson   
1184  Leonardo DiCaprio|Jonah Hill|Margot Robbie|Kyl...    Martin Scorsese   
1192  Chiwetel Ejiofor|Michael Fassbender|Lupita Nyo...      Steve McQueen   
1801  Guy Pearce|Carrie-Anne Moss|Joe Pantoliano|Mar...  Christopher Nolan   

                                                tagline  \
600                                     Why So Serious?   
9              Meet the little voices inside your head.   
34                             Love knows no boundaries   
118   Mankind was born on Earth. It was never meant ...   
125   The true enigma was the man who cracked the code.   
119                         All heroes start somewhere.   
128         You don't know what you've got 'til it's...   
138             A perfect holiday without leaving home.   
370                Your mind is the scene of the crime.   
873   Music was his passion. Survival was his master...   
1082                    The eye of the enemy is moving.   
1184                                EARN. SPEND. PARTY.   
1192    The extraordinary true story of Solomon Northup   
1801                  Some memories are best forgotten.   

                                               overview  runtime  ...  \
600   Batman raises the stakes in his war on crime. ...      152  ...   
9     Growing up can be a bumpy road, and it's no ex...       94  ...   
34    Jack is a young boy of 5 years old who has liv...      117  ...   
118   Interstellar chr

In [82]:
answer_ls.append(1)

# 35. Какие актеры чаще всего снимаются в одном фильме вместе
Варианты ответа:
1. Johnny Depp & Helena Bonham Carter
2. Hugh Jackman & Ian McKellen
3. Vin Diesel & Paul Walker
4. Adam Sandler & Kevin James
5. Daniel Radcliffe & Rupert Grint

In [86]:
answer_ls.append(5)

# 36. У какого из режиссеров выше вероятность выпустить фильм в прибыли? (5 баллов)101
Варианты ответа:
1. Quentin Tarantino
2. Steven Soderbergh
3. Robert Rodriguez
4. Christopher Nolan
5. Clint Eastwood

In [83]:
df_dir = pd.DataFrame(data.director.str.split('|').tolist()).stack().reset_index() 
df_dir.columns =  ['row1', 'row2', 'director']
df_dir['revenue_dir'] = df_dir.row1.apply(lambda x: data.iloc[x]['revenue'])
df_dir['budget'] = df_dir.row1.apply(lambda x: data.iloc[x]['budget'])
dir_over_budget = df_dir.loc[df_dir['revenue_dir'] > df_dir['budget']]
dir_over_budget.groupby(['director'])['revenue_dir'].count().sort_values(ascending = False)

director
Ridley Scott             12
Clint Eastwood           10
Steven Spielberg         10
Steven Soderbergh        10
Tim Burton                9
                         ..
Jean-François Richet     1
Jean-Jacques Annaud       1
Jean-Marc Vallée         1
Jeff Newitt               1
Jonathan van Tulleken     1
Name: revenue_dir, Length: 814, dtype: int64

In [84]:
actor_list = ('Steven Soderbergh,Quentin Tarantino,Robert Rodriguez,Christopher Nolan,Clint Eastwood').\
split(',')
over_budget = data.loc[data['revenue'] > data['budget']]
over_budget = over_budget.director.str.split('|').tolist()
c = Counter()
for i in over_budget:
    for j in i:
        for actor in actor_list:
            if actor in j:
                c[j] += 1
print(c)

Counter({'Clint Eastwood': 10, 'Steven Soderbergh': 10, 'Christopher Nolan': 8, 'Robert Rodriguez': 8, 'Quentin Tarantino': 6})


In [85]:
answer_ls.append(5)

# Submission

In [87]:
len(answer_ls)

36

In [90]:
pd.DataFrame({'Id':range(1,len(answer_ls)+1), 'Answer':answer_ls}, columns=['Id', 'Answer'])

Id  Answer
0    1       4
1    2       2
2    3       3
3    4       2
4    5       1
5    6       5
6    7       2
7    8       1
8    9       4
9   10       5
10  11       3
11  12       1
12  13       3
13  14       4
14  15       5
15  16       1
16  17       3
17  18       3
18  19       2
19  20       1
20  21       4
21  22       2
22  23       3
23  24       1
24  25       5
25  26       1
26  27       4
27  28       2
28  29       5
29  30       2
30  31       5
31  32       5
32  33       3
33  34       1
34  35       5
35  36       5